<a href="https://colab.research.google.com/github/bavcb/association-rule/blob/main/Test_Association_Rule_M02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#*เปลี่ยนจากชื่อสินค้าที่เป็นภาษาจีนเป็นภาษาอังกฤษ และใช้ cat_sub1_name แทน product_name*

# Import Package Library

In [1]:
!pip install mysql-connector-python
!pip install mlxtend
!pip install googletrans==4.0.0-rc1
#!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for googletrans: filename=googletrans-4.0.0rc1-py3-none-any.whl size=17397 sha256=d48e1afcefbe367936092dc92d5e128b69ac37e9d0387cdcee1c944108205e2d
  Stored in directory: /root/.cache/pip/wheels/c0/59/9f/7372f0cf70160fe61b528532e1a7c8498c4becd6bcffb022de
Successfully built googletrans
  Attempting uninstall: googletrans
    Found existing installation: googletrans 3.0.0
    Uninstalling googletrans-3.0.0:
      Successfully uninstalled googletrans-3.0.0


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import mysql.connector
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from mlxtend.preprocessing import TransactionEncoder
import itertools
from googletrans import Translator
#from apyori import apriori

# Import Data

In [3]:
# Establish the connection
conn = mysql.connector.connect(
    host='159.138.255.180',
    user='root',
    password='Vcb168899@',
    database='vcb_ba'
)

# Create a cursor object
cursor = conn.cursor()

# Define SQL query
query = "SELECT * FROM analysis_products"

# Execute the query and fetch data # ระยะเวลาในการดึงข้อมูลนั้นขึ้นอยู่กับขนาดของข้อมูล ยิ่งมีขนาดใหญ่ยิ่งใช้เวลานาน
cursor.execute(query)
rows = cursor.fetchall()

# Get column names
column_names = [desc[0] for desc in cursor.description]


# read sql
#df = pd.read_sql(query,conn)
file = pd.DataFrame(rows,columns=column_names)
# Close the cursor and connection
cursor.close()
conn.close()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [4]:
file.to_csv('data_file.csv', index=False) # change sql to csv # ไฟล์ csv จะถูกเก็บไว้ใน Folder ของไฟล์งานที่ทำอยู่
df = pd.read_csv('data_file.csv') # read csv file

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [5]:
# แสดงข้อมูล
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


id  purchases_id  cars_id       baskets_id      purchase_create  \
0        4129178       2248805  2337512  196678524095069  2021-12-27 10:13:01   
1        4129179       2249918  2338921  196678524101901  2021-12-28 15:01:56   
2        4129180       2249918  2338921  196678524101902  2021-12-28 15:01:56   
3        4129181       2249918  2338921  196678524101903  2021-12-28 15:01:56   
4        4129182       2249918  2338921  196678524101904  2021-12-28 15:01:56   
...          ...           ...      ...              ...                  ...   
2020767  6149945       2816071  3081323  196678528089285  2024-08-07 23:00:02   
2020768  6149946       2816071  3081323  196678528089286  2024-08-07 23:00:02   
2020769  6149947       2816072  3081320  196678528123715  2024-08-07 23:00:10   
2020770  6149948       2816074  3083062  196678528135432  2024-08-08 00:34:28   
2020771  6149949       2816075  3083060  196678528139656  2024-08-08 01:13:45   

            transferred_date chanal_purchase   th_code  level_num  gender  \
0        2022-01-02 12:21:17            1688  TH403375          3       0   
1        2022-01-02 12:21:18            1688  TH174119          4       0   
2        2022-01-02 12:21:18            1688  TH174119          4       0   
3        2022-01-02 12:21:18            1688  TH174119          4       0   
4        2022-01-02 12:21:18            1688  TH174119          4       0   
...                      ...             ...       ...        ...     ...   
2020767  2024-08-08 12:49:38            1688  TH313880          1       0   
2020768  2024-08-08 12:49:38            1688  TH313880          1       0   
2020769  2024-08-08 15:10:33            1688  TH313880          1       0   
2020770  2024-08-08 09:33:55          TAOBAO  TH398422          1       0   
2020771  2024-08-08 11:12:42             NaN   TH23626          1       0   

         ...     cat_sub1  cat_sub1_name   cat_sub2 cat_sub2_name  \
0        ...  127372010.0           大码女装  1046799.0         大码连衣裙   
1        ...    1033273.0           布线产品  1033954.0         线卡、线扣   
2        ...          NaN            NaN        NaN           NaN   
3        ...          NaN            NaN        NaN           NaN   
4        ...          NaN            NaN        NaN           NaN   
...      ...          ...            ...        ...           ...   
2020767  ...          NaN            NaN        NaN           NaN   
2020768  ...          NaN            NaN        NaN           NaN   
2020769  ...  123924007.0             剪刀  1035507.0          其他剪刀   
2020770  ...          NaN            NaN        NaN           NaN   
2020771  ...          NaN            NaN        NaN           NaN   

         product_quantity product_price total_amount country_num  \
0                     1.0         43.06        43.06           1   
1                   200.0          0.55       110.00           1   
2                   100.0          0.80        80.00           1   
3                    50.0          2.60       130.00           1   
4                   200.0          3.80       760.00           1   
...                   ...           ...          ...         ...   
2020767            2500.0          0.43      1075.00           1   
2020768            2500.0          0.53      1325.00           1   
2020769             100.0          9.00       900.00           1   
2020770               1.0         44.98        59.00           1   
2020771               3.0        280.00       840.00           1   

         country_name ck  
0            THAILAND  0  
1            THAILAND  0  
2            THAILAND  0  
3            THAILAND  0  
4            THAILAND  0  
...               ... ..  
2020767      THAILAND  0  
2020768      THAILAND  0  
2020769      THAILAND  0  
2020770      THAILAND  0  
2020771      THAILAND  0  

[2020772 rows x 30 columns]

In [6]:
df.shape # จำนวนข้อมูลทั้งหมด

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


(2020772, 30)

# Clean Data

In [7]:
# ตัดคอลัมน์ที่ไม่ได้ใช้ทิ้ง
data = df[['id','purchases_id','cars_id','baskets_id','transferred_date','th_code','chanal_purchase','po','so','chanal_bill','product_link','product_name','cat_main','cat_main_name','cat_sub1','cat_sub1_name','cat_sub2','cat_sub2_name','product_quantity','product_price','total_amount']]
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


id  purchases_id  cars_id       baskets_id     transferred_date  \
0        4129178       2248805  2337512  196678524095069  2022-01-02 12:21:17   
1        4129179       2249918  2338921  196678524101901  2022-01-02 12:21:18   
2        4129180       2249918  2338921  196678524101902  2022-01-02 12:21:18   
3        4129181       2249918  2338921  196678524101903  2022-01-02 12:21:18   
4        4129182       2249918  2338921  196678524101904  2022-01-02 12:21:18   
...          ...           ...      ...              ...                  ...   
2020767  6149945       2816071  3081323  196678528089285  2024-08-08 12:49:38   
2020768  6149946       2816071  3081323  196678528089286  2024-08-08 12:49:38   
2020769  6149947       2816072  3081320  196678528123715  2024-08-08 15:10:33   
2020770  6149948       2816074  3083062  196678528135432  2024-08-08 09:33:55   
2020771  6149949       2816075  3083060  196678528139656  2024-08-08 11:12:42   

          th_code chanal_purchase           po               so  chanal_bill  \
0        TH403375            1688  21122700096  XLY211226100380            1   
1        TH174119            1688  21122800487  XHY211228100359            1   
2        TH174119            1688  21122800487  XHY211228100359            1   
3        TH174119            1688  21122800487  XHY211228100359            1   
4        TH174119            1688  21122800487  XHY211228100359            1   
...           ...             ...          ...              ...          ...   
2020767  TH313880            1688  24080700634  XHY240805100117            1   
2020768  TH313880            1688  24080700634  XHY240805100117            1   
2020769  TH313880            1688  24080700635  XHY240805100114            1   
2020770  TH398422          TAOBAO  24080800001  XLY240806100836            1   
2020771   TH23626             NaN  24080800002  XLY240806100834            1   

         ...                                       product_name cat_main  \
0        ...                                              大码连衣裙  10166.0   
1        ...                                              线卡、线扣     57.0   
2        ...                                               其他插座      NaN   
3        ...                                               其他插座      NaN   
4        ...                                               船型开关      NaN   
...      ...                                                ...      ...   
2020767  ...                                                NaN      NaN   
2020768  ...                                                NaN      NaN   
2020769  ...                                                NaN     59.0   
2020770  ...  kumikumiหวานลบอายุกดพับขัดแตะบินบินแขนเสื้อเสื...      NaN   
2020771  ...                                                NaN      NaN   

         cat_main_name     cat_sub1  cat_sub1_name   cat_sub2  cat_sub2_name  \
0                   女装  127372010.0           大码女装  1046799.0          大码连衣裙   
1                电子元器件    1033273.0           布线产品  1033954.0          线卡、线扣   
2                  NaN          NaN            NaN        NaN            NaN   
3                  NaN          NaN            NaN        NaN            NaN   
4                  NaN          NaN            NaN        NaN            NaN   
...                ...          ...            ...        ...            ...   
2020767            NaN          NaN            NaN        NaN            NaN   
2020768            NaN          NaN            NaN        NaN            NaN   
2020769          五金、工具  123924007.0             剪刀  1035507.0           其他剪刀   
2020770            NaN          NaN            NaN        NaN            NaN   
2020771            NaN          NaN            NaN        NaN            NaN   

        product_quantity  product_price  total_amount  
0                    1.0          43.06         43.06  
1                  200.0           0.55        110.00  
2                  100.0           0.8

In [8]:
data.isnull().any() # True = NaN

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


id                  False
purchases_id        False
cars_id             False
baskets_id          False
transferred_date    False
th_code             False
chanal_purchase      True
po                  False
so                   True
chanal_bill         False
product_link         True
product_name         True
cat_main             True
cat_main_name        True
cat_sub1             True
cat_sub1_name        True
cat_sub2             True
cat_sub2_name        True
product_quantity    False
product_price       False
total_amount        False
dtype: bool

In [9]:
# ลบค่าว่างทิ้งทั้งหมด
data_drop = data.dropna(subset=['cat_sub1_name'])
data_drop

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


id  purchases_id  cars_id       baskets_id     transferred_date  \
0        4129178       2248805  2337512  196678524095069  2022-01-02 12:21:17   
1        4129179       2249918  2338921  196678524101901  2022-01-02 12:21:18   
6        4129184       2250644  2340008  196678524106956  2022-01-02 12:19:20   
7        4129185       2250661  2339780  196678524105638  2022-01-02 12:19:21   
8        4129186       2250661  2339780  196678524105639  2022-01-02 12:19:21   
...          ...           ...      ...              ...                  ...   
2020671  6149849       2816036  3082463  196678528134311  2024-08-08 15:10:36   
2020682  6149860       2816049  3082422  196678528023446  2024-08-08 12:57:29   
2020683  6149861       2816049  3082422  196678528023447  2024-08-08 12:57:29   
2020685  6149863       2816051  3081970  196678527735731  2024-08-08 12:49:41   
2020769  6149947       2816072  3081320  196678528123715  2024-08-08 15:10:33   

          th_code chanal_purchase           po               so  chanal_bill  \
0        TH403375            1688  21122700096  XLY211226100380            1   
1        TH174119            1688  21122800487  XHY211228100359            1   
6        TH100168            1688  21122900508  XLY211229100501            1   
7        TH059370            1688  21122900525  XLY211229100273            1   
8        TH059370            1688  21122900525  XLY211229100273            1   
...           ...             ...          ...              ...          ...   
2020671  TH365471            1688  24080700599  XHY240806100237            1   
2020682  TH365471            1688  24080700612  XHY240806100196            1   
2020683  TH365471            1688  24080700612  XHY240806100196            1   
2020685  TH365471            1688  24080700614  XHY240805100764            1   
2020769  TH313880            1688  24080700635  XHY240805100114            1   

         ... product_name cat_main  cat_main_name     cat_sub1  cat_sub1_name  \
0        ...        大码连衣裙  10166.0             女装  127372010.0           大码女装   
1        ...        线卡、线扣     57.0          电子元器件    1033273.0           布线产品   
6        ...           餐椅     13.0          家装、建材  123610020.0           餐厅家具   
7        ...    冰袋、冰包、野餐包     15.0           日用百货      10287.0             餐具   
8        ...    冰袋、冰包、野餐包     15.0           日用百货      10287.0             餐具   
...      ...          ...      ...            ...          ...            ...   
2020671  ...          NaN     55.0             橡塑    1031654.0        工业用橡胶制品   
2020682  ...          NaN  10208.0           仪器仪表    1033816.0          工业计时器   
2020683  ...          NaN  10208.0           仪器仪表    1033816.0          工业计时器   
2020685  ...          NaN     59.0          五金、工具  124194006.0           液压工具   
2020769  ...          NaN     59.0          五金、工具  123924007.0             剪刀   

            cat_sub2  cat_sub2_name product_quantity  product_price  \
0          1046799.0          大码连衣裙              1.0          43.06   
1          1033954.0          线卡、线扣            200.0           0.55   
6        121486003.0             餐椅             12.0          53.44   
7          1031741.0      冰袋、冰包、野餐包              2.0          50.00   
8          1031741.0      冰袋、冰包、野餐包              1.0          50.00   
...              ...            ...              ...            ...   
2020671          0.0            NaN            200.0          14.00   
2020682          0.0            NaN            200.0          10.50   
2020683          0.0            NaN            200.0          10.50   
2020685    1036783.0         其他液压工具              6.0         145.00   
2020769    1035507.0           其他剪刀            100.0           9.00   

         total_amount  
0               43.06  
1              110.00  
6              641.28  
7              100.00  
8               50.00  
...               ...  
2020671       2800.00  
2020682       2100.00  
2020683       2100.00  
2020685       

In [10]:
data_drop.shape # จำนวนข้อมูลหลังลบค่าว่างทิ้ง #880845 #1116605

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


(1116605, 21)

In [11]:
data_drop[data_drop['product_quantity'] ==0]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


id  purchases_id  cars_id       baskets_id     transferred_date  \
60       4129238       2251102  2340140  196678524107617  2022-01-02 12:19:56   
78       4129256       2251102  2340140  196678524107635  2022-01-02 12:19:56   
485      4129663       2251536  2342259  196678524117562  2022-01-03 16:40:28   
566      4129744       2251555  2340876  196678524110681  2022-01-03 14:48:47   
727      4129905       2251611  2341115  196678524111837  2022-01-03 14:48:49   
...          ...           ...      ...              ...                  ...   
2012131  6141309       2813844  3078102  196678527871329  2024-08-03 14:56:54   
2012134  6141312       2813844  3078102  196678527871332  2024-08-03 14:56:54   
2012137  6141315       2813844  3078102  196678527871335  2024-08-03 14:56:54   
2012142  6141320       2813844  3078102  196678527871340  2024-08-03 14:56:54   
2014913  6144091       2814581  3081347  196678527397342  2024-08-06 14:27:44   

          th_code chanal_purchase           po               so  chanal_bill  \
60        TH13430            1688  21123000327  XHY211229100633            1   
78        TH13430            1688  21123000327  XHY211229100633            1   
485      TH544422            1688  22010300132  XLY220103100128            1   
566      TH631771            1688  22010300151   LY211230100573            2   
727      TH041271            1688  22010300207  XLY211231100175            2   
...           ...             ...          ...              ...          ...   
2012131  TH416487            1688  24080300173  XLY240731100743            1   
2012134  TH416487            1688  24080300173  XLY240731100743            1   
2012137  TH416487            1688  24080300173  XLY240731100743            1   
2012142  TH416487            1688  24080300173  XLY240731100743            1   
2014913  TH110296            1688  24080500463  XLY240805100141            1   

         ... product_name     cat_main  cat_main_name     cat_sub1  \
60       ...         仿真绿植         67.0          办公、文教  201162103.0   
78       ...         仿真绿植         67.0          办公、文教  201162103.0   
485      ...         小学书包    1042954.0           箱包皮具  126128002.0   
566      ...          休闲裤      10166.0             女装  125748001.0   
727      ...        办公用胶带         67.0          办公、文教       2113.0   
...      ...          ...          ...            ...          ...   
2012131  ...          NaN  127380009.0           运动服饰  121898001.0   
2012134  ...          NaN  127380009.0           运动服饰  121898001.0   
2012137  ...          NaN  127380009.0           运动服饰  121898001.0   
2012142  ...          NaN  127380009.0           运动服饰  121898001.0   
2014913  ...           罩衣       1501.0           母婴用品  122380006.0   

         cat_sub1_name     cat_sub2  cat_sub2_name product_quantity  \
60                仿真园艺    1038172.0           仿真植物              0.0   
78                仿真园艺    1038172.0           仿真植物              0.0   
485                儿童包    1031750.0           小学书包              0.0   
566              女式休闲裤    1031871.0            休闲裤              0.0   
727         装订、胶粘、桌面用品  124746005.0          办公用胶带              0.0   
...                ...          ...            ...              ...   
2012131       运动户外辅助用品    1048301.0         运动帽、头巾              0.0   
2012134       运动户外辅助用品    1048301.0         运动帽、头巾              0.0   
2012137       运动户外辅助用品    1048301.0         运动帽、头巾              0.0   
2012142       运动户外辅助用品    1048301.0         运动帽、头巾              0.0   
2014913       婴幼儿寝居/服饰  124958003.0             罩衣              0.0   

         product_price  total_amount  
60                3.95           0.0  
78                5.50           0.0  
485             112.00           0.0  
566              15.99           0.0  
727               2.70           0.0  
...                ...           ...  
2012131           7.50           0.0  
2012134           7.50           0.0  
2012137           7.50           0.

In [12]:
data_drop[data_drop['total_amount'] ==0]


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


id  purchases_id  cars_id       baskets_id     transferred_date  \
60       4129238       2251102  2340140  196678524107617  2022-01-02 12:19:56   
78       4129256       2251102  2340140  196678524107635  2022-01-02 12:19:56   
485      4129663       2251536  2342259  196678524117562  2022-01-03 16:40:28   
566      4129744       2251555  2340876  196678524110681  2022-01-03 14:48:47   
727      4129905       2251611  2341115  196678524111837  2022-01-03 14:48:49   
...          ...           ...      ...              ...                  ...   
2012131  6141309       2813844  3078102  196678527871329  2024-08-03 14:56:54   
2012134  6141312       2813844  3078102  196678527871332  2024-08-03 14:56:54   
2012137  6141315       2813844  3078102  196678527871335  2024-08-03 14:56:54   
2012142  6141320       2813844  3078102  196678527871340  2024-08-03 14:56:54   
2014913  6144091       2814581  3081347  196678527397342  2024-08-06 14:27:44   

          th_code chanal_purchase           po               so  chanal_bill  \
60        TH13430            1688  21123000327  XHY211229100633            1   
78        TH13430            1688  21123000327  XHY211229100633            1   
485      TH544422            1688  22010300132  XLY220103100128            1   
566      TH631771            1688  22010300151   LY211230100573            2   
727      TH041271            1688  22010300207  XLY211231100175            2   
...           ...             ...          ...              ...          ...   
2012131  TH416487            1688  24080300173  XLY240731100743            1   
2012134  TH416487            1688  24080300173  XLY240731100743            1   
2012137  TH416487            1688  24080300173  XLY240731100743            1   
2012142  TH416487            1688  24080300173  XLY240731100743            1   
2014913  TH110296            1688  24080500463  XLY240805100141            1   

         ... product_name     cat_main  cat_main_name     cat_sub1  \
60       ...         仿真绿植         67.0          办公、文教  201162103.0   
78       ...         仿真绿植         67.0          办公、文教  201162103.0   
485      ...         小学书包    1042954.0           箱包皮具  126128002.0   
566      ...          休闲裤      10166.0             女装  125748001.0   
727      ...        办公用胶带         67.0          办公、文教       2113.0   
...      ...          ...          ...            ...          ...   
2012131  ...          NaN  127380009.0           运动服饰  121898001.0   
2012134  ...          NaN  127380009.0           运动服饰  121898001.0   
2012137  ...          NaN  127380009.0           运动服饰  121898001.0   
2012142  ...          NaN  127380009.0           运动服饰  121898001.0   
2014913  ...           罩衣       1501.0           母婴用品  122380006.0   

         cat_sub1_name     cat_sub2  cat_sub2_name product_quantity  \
60                仿真园艺    1038172.0           仿真植物              0.0   
78                仿真园艺    1038172.0           仿真植物              0.0   
485                儿童包    1031750.0           小学书包              0.0   
566              女式休闲裤    1031871.0            休闲裤              0.0   
727         装订、胶粘、桌面用品  124746005.0          办公用胶带              0.0   
...                ...          ...            ...              ...   
2012131       运动户外辅助用品    1048301.0         运动帽、头巾              0.0   
2012134       运动户外辅助用品    1048301.0         运动帽、头巾              0.0   
2012137       运动户外辅助用品    1048301.0         运动帽、头巾              0.0   
2012142       运动户外辅助用品    1048301.0         运动帽、头巾              0.0   
2014913       婴幼儿寝居/服饰  124958003.0             罩衣              0.0   

         product_price  total_amount  
60                3.95           0.0  
78                5.50           0.0  
485             112.00           0.0  
566              15.99           0.0  
727               2.70           0.0  
...                ...           ...  
2012131           7.50           0.0  
2012134           7.50           0.0  
2012137           7.50           0.

In [13]:
# เอา product_quantity & total_amount ที่มีค่าเป็น 0 ออก
dataset = data_drop[(data_drop.product_quantity > 0) & (data_drop.total_amount >0)]
dataset

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


id  purchases_id  cars_id       baskets_id     transferred_date  \
0        4129178       2248805  2337512  196678524095069  2022-01-02 12:21:17   
1        4129179       2249918  2338921  196678524101901  2022-01-02 12:21:18   
6        4129184       2250644  2340008  196678524106956  2022-01-02 12:19:20   
7        4129185       2250661  2339780  196678524105638  2022-01-02 12:19:21   
8        4129186       2250661  2339780  196678524105639  2022-01-02 12:19:21   
...          ...           ...      ...              ...                  ...   
2020671  6149849       2816036  3082463  196678528134311  2024-08-08 15:10:36   
2020682  6149860       2816049  3082422  196678528023446  2024-08-08 12:57:29   
2020683  6149861       2816049  3082422  196678528023447  2024-08-08 12:57:29   
2020685  6149863       2816051  3081970  196678527735731  2024-08-08 12:49:41   
2020769  6149947       2816072  3081320  196678528123715  2024-08-08 15:10:33   

          th_code chanal_purchase           po               so  chanal_bill  \
0        TH403375            1688  21122700096  XLY211226100380            1   
1        TH174119            1688  21122800487  XHY211228100359            1   
6        TH100168            1688  21122900508  XLY211229100501            1   
7        TH059370            1688  21122900525  XLY211229100273            1   
8        TH059370            1688  21122900525  XLY211229100273            1   
...           ...             ...          ...              ...          ...   
2020671  TH365471            1688  24080700599  XHY240806100237            1   
2020682  TH365471            1688  24080700612  XHY240806100196            1   
2020683  TH365471            1688  24080700612  XHY240806100196            1   
2020685  TH365471            1688  24080700614  XHY240805100764            1   
2020769  TH313880            1688  24080700635  XHY240805100114            1   

         ... product_name cat_main  cat_main_name     cat_sub1  cat_sub1_name  \
0        ...        大码连衣裙  10166.0             女装  127372010.0           大码女装   
1        ...        线卡、线扣     57.0          电子元器件    1033273.0           布线产品   
6        ...           餐椅     13.0          家装、建材  123610020.0           餐厅家具   
7        ...    冰袋、冰包、野餐包     15.0           日用百货      10287.0             餐具   
8        ...    冰袋、冰包、野餐包     15.0           日用百货      10287.0             餐具   
...      ...          ...      ...            ...          ...            ...   
2020671  ...          NaN     55.0             橡塑    1031654.0        工业用橡胶制品   
2020682  ...          NaN  10208.0           仪器仪表    1033816.0          工业计时器   
2020683  ...          NaN  10208.0           仪器仪表    1033816.0          工业计时器   
2020685  ...          NaN     59.0          五金、工具  124194006.0           液压工具   
2020769  ...          NaN     59.0          五金、工具  123924007.0             剪刀   

            cat_sub2  cat_sub2_name product_quantity  product_price  \
0          1046799.0          大码连衣裙              1.0          43.06   
1          1033954.0          线卡、线扣            200.0           0.55   
6        121486003.0             餐椅             12.0          53.44   
7          1031741.0      冰袋、冰包、野餐包              2.0          50.00   
8          1031741.0      冰袋、冰包、野餐包              1.0          50.00   
...              ...            ...              ...            ...   
2020671          0.0            NaN            200.0          14.00   
2020682          0.0            NaN            200.0          10.50   
2020683          0.0            NaN            200.0          10.50   
2020685    1036783.0         其他液压工具              6.0         145.00   
2020769    1035507.0           其他剪刀            100.0           9.00   

         total_amount  
0               43.06  
1              110.00  
6              641.28  
7              100.00  
8               50.00  
...               ...  
2020671       2800.00  
2020682       2100.00  
2020683       2100.00  
2020685       

## แปลภาษาจีน ---> อังกฤษ

In [14]:
translator = Translator()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/httpx/_config.py:156: DeprecationWarning: ssl.PROTOCOL_TLS is deprecated
  context = ssl.SSLContext(ssl.PROTOCOL_TLS)
/usr/local/lib/python3.10/dist-packages/httpx/_config.py:159: DeprecationWarning: ssl.OP_NO_SSL*/ssl.OP_NO_TLS* options are deprecated
  context.options |= ssl.OP_NO_TLSv1
/usr/local/lib/python3.10/dist-packages/httpx/_config.py:160: DeprecationWarning: ssl.OP_NO_SSL*/ssl.OP_NO_TLS* options are deprecated
  context.options |= ssl.OP_NO_TLSv1_1


In [15]:
# ฟังก์ชันในการแปลข้อความ
def translate_text(text, src_lang='auto', dest_lang='en'):
    try:
        return translator.translate(text, src=src_lang, dest=dest_lang).text
    except Exception as e:
        return text  # ถ้าแปลไม่ได้ให้คืนค่าเดิม

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [16]:
translation_cache = {} # ใช้ cache ในการช่วยลดระยะเวลาในการแปลข้อความ โดยถ้าเจอข้อความซ้ำกันกับที่แปลไปแล้วให้ดึงคำแปลนั้นมาใช้แทนการแปลภาษา

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [17]:
def translate_text_with_cache(text, dest_lang='en'):
    if text in translation_cache:
        return translation_cache[text]
    else:
        translated = translate_text(text, dest_lang=dest_lang)
        translation_cache[text] = translated
        return translated

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [18]:
dataset['cat_sub1_name_en'] = dataset['cat_sub1_name'].apply(lambda x: translate_text_with_cache(x, dest_lang='en'))

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
<ipython-input-18-f06aa81dcdf8>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset['cat_sub1_name_en'] = dataset['cat_sub1_name'].apply(lambda x: translate_text_with_cache(x, dest_lang='en'))


In [19]:
#dataset['cat_sub1_name_en'] = dataset['cat_sub1_name'].apply(lambda x: translate_text(x, dest_lang='en')) # ใช้เวลานานมากประมาณ 70 นาที

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [20]:
pd.set_option('display.max_columns', None)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [21]:
dataset

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


id  purchases_id  cars_id       baskets_id     transferred_date  \
0        4129178       2248805  2337512  196678524095069  2022-01-02 12:21:17   
1        4129179       2249918  2338921  196678524101901  2022-01-02 12:21:18   
6        4129184       2250644  2340008  196678524106956  2022-01-02 12:19:20   
7        4129185       2250661  2339780  196678524105638  2022-01-02 12:19:21   
8        4129186       2250661  2339780  196678524105639  2022-01-02 12:19:21   
...          ...           ...      ...              ...                  ...   
2020671  6149849       2816036  3082463  196678528134311  2024-08-08 15:10:36   
2020682  6149860       2816049  3082422  196678528023446  2024-08-08 12:57:29   
2020683  6149861       2816049  3082422  196678528023447  2024-08-08 12:57:29   
2020685  6149863       2816051  3081970  196678527735731  2024-08-08 12:49:41   
2020769  6149947       2816072  3081320  196678528123715  2024-08-08 15:10:33   

          th_code chanal_purchase           po               so  chanal_bill  \
0        TH403375            1688  21122700096  XLY211226100380            1   
1        TH174119            1688  21122800487  XHY211228100359            1   
6        TH100168            1688  21122900508  XLY211229100501            1   
7        TH059370            1688  21122900525  XLY211229100273            1   
8        TH059370            1688  21122900525  XLY211229100273            1   
...           ...             ...          ...              ...          ...   
2020671  TH365471            1688  24080700599  XHY240806100237            1   
2020682  TH365471            1688  24080700612  XHY240806100196            1   
2020683  TH365471            1688  24080700612  XHY240806100196            1   
2020685  TH365471            1688  24080700614  XHY240805100764            1   
2020769  TH313880            1688  24080700635  XHY240805100114            1   

                                            product_link product_name  \
0        https://detail.1688.com/offer/651330149311.html        大码连衣裙   
1        https://detail.1688.com/offer/555962625187.html        线卡、线扣   
6        https://detail.1688.com/offer/658662964846.html           餐椅   
7          https://detail.1688.com/offer/1295274089.html    冰袋、冰包、野餐包   
8          https://detail.1688.com/offer/1295274089.html    冰袋、冰包、野餐包   
...                                                  ...          ...   
2020671  https://detail.1688.com/offer/564571594343.html          NaN   
2020682  https://detail.1688.com/offer/574076029717.html          NaN   
2020683  https://detail.1688.com/offer/573932012067.html          NaN   
2020685  https://detail.1688.com/offer/644257273431.html          NaN   
2020769  https://detail.1688.com/offer/664644896562.html          NaN   

         cat_main cat_main_name     cat_sub1 cat_sub1_name     cat_sub2  \
0         10166.0            女装  127372010.0          大码女装    1046799.0   
1            57.0         电子元器件    1033273.0          布线产品    1033954.0   
6            13.0         家装、建材  123610020.0          餐厅家具  121486003.0   
7            15.0          日用百货      10287.0            餐具    1031741.0   
8            15.0          日用百货      10287.0            餐具    1031741.0   
...           ...           ...          ...           ...          ...   
2020671      55.0            橡塑    1031654.0       工业用橡胶制品          0.0   
2020682   10208.0          仪器仪表    1033816.0         工业计时器          0.0   
2020683   10208.0          仪器仪表    1033816.0         工业计时器          0.0   
2020685      59.0         五金、工具  124194006.0          液压工具    1036783.0   
2020769      59.0         五金、工具  123924007.0            剪刀    1035507.0   

        cat_sub2_name  product_quantity  product_price  total_amount  \
0               大码连衣裙               1.0          43.06         43.06   
1               线卡、线扣             200.0           0.55        110.00   
6                  餐椅              12.0          53.44        641.28   
7           冰袋

# Data Preprocessing

**คอลัมน์ที่นำมาใช้งาน**

  *purchase_id == รหัสคำสั่งซื้อ*

  *product_name == ชื่่อสินค้า*

  *th_code == รหัสลูกค้า*

In [22]:
# ใน 1 การสั่งซื้อลูกค้าซื้อสินค้าไปจำนวนกี่รายการ
purchase = dataset.groupby('purchases_id').count().sort_values(by=['cat_sub1_name_en'],ascending=False)
purchase.iloc[:,-1:].reset_index()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


purchases_id  cat_sub1_name_en
0            2691827               144
1            2691925               135
2            2741069                88
3            2691969                84
4            2691816                81
...              ...               ...
209310       2613030                 1
209311       2613029                 1
209312       2434937                 1
209313       2613025                 1
209314       2816072                 1

[209315 rows x 2 columns]

In [23]:
# ใน 1 การสั่งซื้อลูกค้าซื้อสินค้าไปจำนวนกี่รายการ โดยที่สินค้าไม่ซ้ำ
product_purchase = dataset.groupby('purchases_id')['cat_sub1_name_en'].nunique().sort_values(ascending=False)
product_purchase.reset_index()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


purchases_id  cat_sub1_name_en
0            2437869                20
1            2438748                18
2            2426847                17
3            2262953                12
4            2438512                12
...              ...               ...
209310       2433194                 1
209311       2433195                 1
209312       2433197                 1
209313       2433198                 1
209314       2816072                 1

[209315 rows x 2 columns]

In [24]:
# ลูกค้า 1 คนสั่งซื้อสินค้าไปกี่รายงาน
th_code = dataset.groupby('th_code').count().sort_values(by=['cat_sub1_name_en'],ascending=False)
th_code.iloc[:,-1:].reset_index()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


th_code  cat_sub1_name_en
0      TH509912             15559
1      TH110296             11236
2      TH124928              9902
3      TH306141              9662
4      TH328224              8940
...         ...               ...
12649  TH784937                 1
12650  TH784997                 1
12651  TH374262                 1
12652  TH785399                 1
12653  TH636592                 1

[12654 rows x 2 columns]

In [25]:
#  ลูกค้า 1 คนสั่งซื้อสินค้าไปกี่รายงาน โดยที่สินค้าไม่ซ้ำ
th_code_purchase = dataset.groupby('th_code')['cat_sub1_name_en'].nunique().sort_values(ascending=False)
th_code_purchase.reset_index()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


th_code  cat_sub1_name_en
0       TH00666               149
1      TH365471               141
2      TH114023               127
3      TH652464               112
4      TH069347               109
...         ...               ...
12649  TH633952                 1
12650  TH633230                 1
12651  TH633219                 1
12652  TH632720                 1
12653  TH636592                 1

[12654 rows x 2 columns]

In [26]:
# ลูกค้า 1 คนสั่งซื้อสินค้าไปกี่รายงาน
products = dataset.groupby('cat_sub1_name_en').count().sort_values(by=['purchases_id'],ascending=False)
products = products.iloc[:,1:2].reset_index()
products

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


cat_sub1_name_en  purchases_id
0                            Jewelry         29583
1        Hair accessories, headdress         27477
2                      Women's shoes         26921
3                             Handle         25055
4                    Children's suit         24962
..                               ...           ...
833              Synthetic materials             1
834                   Rubber machine             1
835  Clothing accessories processing             1
836                              Gas             1
837                       3D printer             1

[838 rows x 2 columns]

In [27]:
transaction = dataset.groupby('th_code')['cat_sub1_name_en'].apply(lambda x: list(set(x))).reset_index()
transaction

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


th_code                                   cat_sub1_name_en
0      TH000036                                           [Garden]
1      TH000041  [adhesive tape, Learning stationery, Doll, dol...
2      TH001489                                 [Special lighting]
3       TH00469  [Jewelry box, bag, bag, box, Women's T -shirt,...
4       TH00577                                 [Living appliance]
...         ...                                                ...
12649  TH979509                                     [Kitchen tool]
12650  TH979567                                    [Women's shoes]
12651  TH979942       [Wind blocking, shading, rainproofing tools]
12652  TH980538          [Pajamas, home clothes, robes, bathrobes]
12653   TH99999  [Cats and supplies, knife, Dogs and supplies, ...

[12654 rows x 2 columns]

In [28]:
#set_list = dataset.groupby(['purchases_id', 'product_name'])['product_name'].count().unstack().reset_index().fillna(0).set_index('purchases_id')
#set_list

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [29]:
transaction_list = transaction['cat_sub1_name_en'].tolist()
te = TransactionEncoder()
te_ary = te.fit(transaction_list).transform(transaction_list)
transaction_set = pd.DataFrame(te_ary, columns=te.columns_)
transaction_set

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


3D printer  3D printing wire  Abrasion  \
0           False             False     False   
1           False             False     False   
2           False             False     False   
3           False             False     False   
4           False             False     False   
...           ...               ...       ...   
12649       False             False     False   
12650       False             False     False   
12651       False             False     False   
12652       False             False     False   
12653       False             False     False   

       Access control attendance equipment and system  \
0                                               False   
1                                               False   
2                                               False   
3                                               False   
4                                               False   
...                                               ...   
12649                                           False   
12650                                           False   
12651                                           False   
12652                                           False   
12653                                           False   

       Administrative supplies  Adsorbent  Advertising shirt  \
0                        False      False              False   
1                        False      False              False   
2                        False      False              False   
3                        False      False              False   
4                        False      False              False   
...                        ...        ...                ...   
12649                    False      False              False   
12650                    False      False              False   
12651                    False      False              False   
12652                    False      False              False   
12653                    False      False              False   

       Advertising, creative towel  Advertising, exhibition equipment  \
0                            False                              False   
1                            False                              False   
2                            False                              False   
3                            False                              False   
4                            False                              False   
...                            ...                                ...   
12649                        False                              False   
12650                        False                              False   
12651                        False                              False   
12652                        False                              False   
12653                        False                              False   

       Affairs of animal husbandry  Agricultural machinery  Agricultural tool  \
0                            False                   False              False   
1                            False                   False              False   
2                            False                   False              False   
3                            False                   False              False   
4                            False                   False              False   
...                            ...                     ...                ...   
12649                        False                   False              False   
12650                        False                   False              False   
12651                        False                   False              False   
12652                        False                   False              False   
12653                        False                   False              False   

       Agricultural utensil  Air purification equipment  Alarm  \
0                     False                     

In [30]:
transaction_set.isnull().any()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


3D printer                                        False
3D printing wire                                  False
Abrasion                                          False
Access control attendance equipment and system    False
Administrative supplies                           False
                                                  ...  
tweezers                                          False
valve                                             False
vegetable                                         False
wallet                                            False
wig                                               False
Length: 838, dtype: bool

In [31]:
filter_test = dataset[dataset['cat_sub1_name_en'] == 'tableware']
filter_test.iloc[:,5:6].reset_index()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


index   th_code
0           7  TH059370
1           8  TH059370
2         450  TH035563
3         451  TH035563
4        1485  TH633952
...       ...       ...
6845  1921005  TH685866
6846  1933040  TH457366
6847  1947766  TH062263
6848  1947767  TH062263
6849  1947768  TH062263

[6850 rows x 2 columns]

In [32]:
test = filter_test.groupby('cat_sub1_name_en')['th_code'].apply(lambda x: ','.join(set(x))).reset_index() \
    .assign(th_code=lambda df: df['th_code'].str.split(',')) \
    .explode('th_code').reset_index(drop=True)
test

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


cat_sub1_name_en   th_code
0          tableware  TH417120
1          tableware  TH319609
2          tableware  TH097198
3          tableware  TH870783
4          tableware  TH636955
..               ...       ...
332        tableware  TH748660
333        tableware  TH348965
334        tableware  TH784393
335        tableware  TH128867
336        tableware  TH648734

[337 rows x 2 columns]

# Model

In [33]:
frequent_itemsets = apriori(transaction_set,min_support = 0.01, use_colnames=True) ## 0.01% ของรายการสินค้าทั้งหมดที่มีคนซื้อมากกว่า 1 %
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                                  itemsets
0    0.011854             (Apple dedicated accessories)
1    0.011538      (Art, calligraphy, drawing supplies)
2    0.028370             (Atmosphere, layout supplies)
3    0.011775                                  (Baking)
4    0.016991                       (Bathroom cleaning)
..        ...                                       ...
113  0.011617  (Women's T -shirt, Women's casual pants)
114  0.010985       (Women's T -shirt, Women's sweater)
115  0.011617                 (Women's T -shirt, dress)
116  0.010511                  (dress, Women's sweater)
117  0.011617                   (Handle, earring, Item)

[118 rows x 2 columns]

In [34]:
#for result in frequent_itemsets:
  #print(result)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [35]:
rules = association_rules(frequent_itemsets, metric="support", min_threshold=0.01)
rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                      consequents  \
0                         (Jewelry)                          (Craft)   
1                           (Craft)                        (Jewelry)   
2                           (dress)            (Fashion casual suit)   
3             (Fashion casual suit)                          (dress)   
4                         (Jewelry)    (Hair accessories, headdress)   
5     (Hair accessories, headdress)                        (Jewelry)   
6                          (Handle)                           (Item)   
7                            (Item)                         (Handle)   
8                         (Jewelry)                         (Handle)   
9                          (Handle)                        (Jewelry)   
10                         (Handle)                        (earring)   
11                        (earring)                         (Handle)   
12                        (Jewelry)                           (Item)   
13                           (Item)                        (Jewelry)   
14                        (earring)                           (Item)   
15                           (Item)                        (earring)   
16                        (Jewelry)                        (earring)   
17                        (earring)                        (Jewelry)   
18            (Learning stationery)                   (Writing tool)   
19                   (Writing tool)            (Learning stationery)   
20  (Small vest/suspender/tube top)               (Women's T -shirt)   
21               (Women's T -shirt)  (Small vest/suspender/tube top)   
22  (Small vest/suspender/tube top)                (Women's sweater)   
23                (Women's sweater)  (Small vest/suspender/tube top)   
24  (Small vest/suspender/tube top)                          (dress)   
25                          (dress)  (Small vest/suspender/tube top)   
26               (Women's T -shirt)           (Women's casual pants)   
27           (Women's casual pants)               (Women's T -shirt)   
28               (Women's T -shirt)                (Women's sweater)   
29                (Women's sweater)               (Women's T -shirt)   
30               (Women's T -shirt)                          (dress)   
31                          (dress)               (Women's T -shirt)   
32                          (dress)                (Women's sweater)   
33                (Women's sweater)                          (dress)   
34                (Handle, earring)                           (Item)   
35                   (Handle, Item)                        (earring)   
36                  (earring, Item)                         (Handle)   
37                         (Handle)                  (earring, Item)   
38                        (earring)                   (Handle, Item)   
39                           (Item)                (Handle, earring)   

    antecedent support  consequent support   support  confidence       lift  \
0             0.051683            0.042516  0.012249    0.237003   5.574418   
1             0.042516            0.051683  0.012249    0.288104   5.574418   
2             0.038486            0.020468  0.012328    0.320329  15.650337   
3             0.020468            0.038486  0.012328    0.602317  15.650337   
4             0.051683            0.060929  0.011459    0.221713   3.638846   
5             0.060929            0.051683  0.011459    0.188067   3.638846   
6             0.037538            0.031769  0.018492    0.492632  15.506866   
7             0.031769            0.037538  0.018492    0.582090  15.506866   
8             0.051683            0.037538  0.012565    0.243119   6.476697   
9             0.037538            0.051683  0.012565    0.334737   6.476697   
10            0.037538            0.028608  0.015015    0.400000  13.982320   
11            0.028608            0.037538  0.015015    0.524862  13.982320   
12            0.051683            0.031769  

**ถ้าลูกค้าซื้อ 发圈 จะซื้อ 发夹 ไปด้วยที่ค่า support = 1.3%, confidance = 55.5 %, lift = 14.4**

**-----------------------------------------------------------------------------------------------------------------------**